internal/enricher/geolocation.go

```diff
@@ -0,0 +1,185 @@
+package enricher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/filemanager"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	"github.com/rudderlabs/rudder-server/services/geolocation"
+	"github.com/rudderlabs/rudder-server/utils/types"
+)
+
+type GeoDBProvider interface {
+	GetDB(string, string) error
+}
+
+type Geolocation struct {
+	City     string `json:"city"`
+	Country  string `json:"country"`
+	Region   string `json:"region"`
+	Postal   string `json:"postal"`
+	Location string `json:"location"`
+	Timezone string `json:"timezone"`
+}
+
+type geoEnricher struct {
+	fetcher geolocation.GeoFetcher
+	logger  logger.Logger
+	stats   stats.Stats
+}
+
+func NewGeoEnricher(
+	dbProvider GeoDBProvider,
+	config *config.Config,
+	log logger.Logger,
+	statClient stats.Stats,
+) (PipelineEnricher, error) {
+	upstreamDBKey := config.GetString("Geolocation.db.key", "geolite2City.mmdb")
+	downloadPath := config.GetString("Geolocation.db.downloadPath", "geolite2City.mmdb")
+
+	err := dbProvider.GetDB(upstreamDBKey, downloadPath)
+	if err != nil {
+		return nil, fmt.Errorf("getting db from upstream: %w", err)
+	}
+
+	fetcher, err := geolocation.NewMaxmindDBReader(downloadPath)
+	if err != nil {
+		return nil, fmt.Errorf("creating new instance of maxmind's geolocation fetcher: %w", err)
+	}
+
+	return &geoEnricher{
+		fetcher: fetcher,
+		stats:   statClient,
+		logger:  log.Child("geolocation"),
+	}, nil
+}
+
+// Enrich function runs on a request of GatewayBatchRequest which contains
+// multiple singular events from a source. The enrich function augments the
+// geolocation information per event based on IP address.
+func (e *geoEnricher) Enrich(sourceId string, request *types.GatewayBatchRequest) error {
+	e.logger.Debugw("received a call to enrich gateway events for source", "sourceId", sourceId)
+
+	if request.RequestIP == "" {
+		e.stats.NewTaggedStat("proc_geo_enrincher_empty_ip", stats.CountType, stats.Tags{
```

Comment:
What if we use a single stat with multiple tags: 

stat name:
`proc_geo_enrincher_request`
tags:
* `sourceId=sourceID`
* `error` 
   * `empty_ip`
   * `lookup_failed`
   * `` (empty)

This way we can easily compare the percentage of empty or lookup failed

Reply:
Good point, I'll make the change.